PyInstaller/loader/pyimod03_importers.py

```diff
@@ -491,6 +491,119 @@ def get_filename(self, fullname):
+
+        Finders must return ModuleSpec objects when find_spec() is called. This new method replaces find_module()
+        and find_loader() (in the PathEntryFinder case). If a loader does not have find_spec(), find_module() and
+        find_loader() are used instead, for backward-compatibility.
+        """
+
+        filename = None  # None means - no module found in this importer.
+
+        if fullname in self.toc:
+            # Tell the import machinery to use self.load_module() to load the module.
+            filename = fullname
+            trace("import %s # PyInstaller PYZ", fullname)
+        elif path is not None:
+            # Try to handle module.__path__ modifications by the modules themselves
+            # Reverse the fake __path__ we added to the package module to a
+            # dotted module name and add the tail module from fullname onto that
+            # to synthesize a new fullname
+            modname = fullname.rsplit('.')[-1]
+
+            for p in path:
+                p = p[SYS_PREFIXLEN:]
+                parts = p.split(pyi_os_path.os_sep)
+                if not parts: continue
+                if not parts[0]:
+                    parts = parts[1:]
+                parts.append(modname)
+                path = ".".join(parts)
+                if path in self.toc:
+                    filename = path
+                    trace("import %s as %s # PyInstaller PYZ (__path__ override: %s)",
+                          path, fullname, p)
+                    break
+
+        if filename is None:
+            trace("# %s not found in PYZ", fullname)
+            return None
+
+        is_pkg, bytecode = self._pyz_archive.extract(filename)
+        if is_pkg:
+            origin = pyi_os_path.os_path_join(pyi_os_path.os_path_join(
+                SYS_PREFIX, filename.replace('.', pyi_os_path.os_sep)), '__init__.pyc')
+        else:
+            origin = pyi_os_path.os_path_join(SYS_PREFIX, filename.replace('.', pyi_os_path.os_sep) + '.pyc')
+
+        if not hasattr(self, 'module_code'):
+            self.module_code = {}
+
+        self.module_code[fullname] = bytecode
+        return _frozen_importlib.ModuleSpec(fullname, self, is_package=is_pkg, origin=origin)
+
+    def create_module(self, spec):
+        """
+        PEP-451 loader.create_module() method for the ``sys.meta_path`` hook.
+
+        Loaders may also implement create_module() that will return a new module to exec. It may return None to
+        indicate that the default module creation code should be used. One use case, though atypical, for
+        create_module() is to provide a module that is a subclass of the builtin module type. Most loaders will not
+        need to implement create_module(),
+
+        create_module() should properly handle the case where it is called more than once for the same spec/module.
+        This may include returning None or raising ImportError.
+        """
+        return imp_new_module(spec.name)
```

Comment:
According to core developer Brett Cannon:

> I continue to be bothered by how we designed importlib.abc.Loader.create_module(). I don't like that it's optional and I don't like that it can return None.

The resolution to this was:

> I don't think we really need to say anything. If people want default results, simply return None (which is handled for them by importlib.abc.Loader). The only thing changing here is that the method will now be required instead of optional.

See https://bugs.python.org/issue23014 for more information.